src/basicfuns.jl

```diff
@@ -152,9 +152,65 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(_x::Real)
+    x = float(_x)
+    x0, x1, x2 = _log1pexp_thresholds(x)
+    if x < x0
+        return exp(x)
+    elseif x < x1
+        return log1p(exp(x))
+    elseif x < x2
+        return x + exp(-x)
+    else
+        return x
+    end
+end
+
+#= The precision of BigFloat cannot be computed from the type only and computing
+thresholds is slow. Therefore prefer version without thresholds in this case. =#
+function log1pexp(x::BigFloat)
+    if x > 0
+        return x + log1p(exp(-x))
+    else
+        return log1p(exp(x))
+    end
+end
+
+#=
+Returns thresholds x0, x1, x2 such that:
+
+    * log1pexp(x) ≈ exp(x) for x ≤ x0
+    * log1pexp(x) ≈ log1p(exp(x)) for x0 < x ≤ x1
+    * log1pexp(x) ≈ x + exp(-x) for x1 < x ≤ x2
+    * log1pexp(x) ≈ x for x > x2
+
+where the tolerances of the approximations are on the order of eps(typeof(x)).
+For types for which `precision(x)` depends only on the type of `x`, the compiler
+should optimize away all computations done here.
+=#
+@inline function _log1pexp_thresholds(x::Real)
+    prec = precision(x)
+    logtwo = oftype(x, IrrationalConstants.logtwo)
+    x0 = -prec * logtwo
+    x1 = (prec - 1) * logtwo / 2
+    x2 = -x0 - log(-x0) * (1 + 1 / x0) # approximate root of e^-x == x * ϵ/2 via asymptotics of Lambert's W function
+    return (x0, x1, x2)
+end
+
+#= For Float64, Float32 we can hard-code the thresholds to make absolutely sure they are not
+recompued each time. Also, _log1pexp_thresholds is not completely elided by the
+compiler in Julia 1.0 / 1.6 which LogExpFunctions intends to support.
+For Float64 we use the same exact thresholds given by Maechler 2012,
+since these were the ones used before _log1pexp_thresholds was introduced (which outputs
+close but not identical thresholds), to reduce any (small) risk of breakage.
+For Float32 and Float16 we use truncated versions of the output of _log1pexp_thresholds. =#
```

Comment:
Maybe it would be interesting to add the precision-based values as a comment?

Reply:
Not sure what you mean?

Reply:
I mean the values you would get from the non-hardcoded version. Then it would be easier to see how it's rounded.

Reply:
Ah yes, ok.